### DISCUSSION TITLE
How to do service workers in parcel 2?

### DISCUSSION THREAD

**Discussion by m-lautenbach** [parcel-bundler/parcel#6098] [Q&A]
Hey 👋 

I'm running into problems with service workers in parcel 2.
I got mine running, but with massive workarounds.

Problem 1: Service worker doesn't execute
When I do `navigator.serviceWorker.register(new URL('./sw.js', import.meta.url))` and debug, I can see that the service worker gets compiled, but never run (no `mainEntry` in generated file).
My workaround is `navigator.serviceWorker.register(new URL('rawjs:./sw.js', import.meta.url))`

Problem 2: service worker parsing fails because `export` is undefined
Parcel generates an `export` for the file, which is apparently not supported by service workers.
I didn't find a way to control that, so I'm currently deploying the `yarn parcel serve` build, which doesn't have that problem, but it's not ideal 😬 

Before I reduce my project to minimal code, is there maybe a guide for service workers with parcel 2?

I combed the internet already, but didn't find anything conclusive (much is for parcel 1 or depends on Workbox, which I'd like to avoid, as my service workers is quite simple).

**Comment by mischnic** [comment]
I'm currently getting `Uncaught ReferenceError: parcelRequire427e is not defined` with just

index.html: `<script src="index.js"></script>`
index.js:  `navigator.serviceWorker.register(new URL("./sw.js", import.meta.url));`
sw.js: `console.log("sw")`

Interestingly, `navigator.serviceWorker.register("./sw.js");` works correctly. 

Is that what you're seeing as well?

**Reply by m-lautenbach** [reply]
> Interestingly, navigator.serviceWorker.register("./sw.js"); works correctly.

Surprisingly, yes (I thought I tried that before, but maybe not in that combination).
I'm on nightly, by the way.

Sounds to me like with `navigator.serviceWorker.register(new URL('rawjs:sw.js', import.meta.url))` the service worker context is not detected? (based on my limited understanding of how parcel works)

**Reply by m-lautenbach** [reply]
Actually, I still get `Uncaught SyntaxError: Unexpected token 'export'` if I build and run (instead of `parcel serve`).

But the service worker does execute now, without using `raw: rawjs:sw.js` 👍 

Which enables me to bring back my offline page.
Thanks for that :)
